### DISCUSSION TITLE
ESP32 reset - Stack canary watchpoint triggered (Modbus01TCP)

### DISCUSSION THREAD

**Discussion by diegostf** [eModbus/eModbus#272] [Q&A] [answered]
I'm testing Modbus TCP on ESP32 (ModbusClientTCP). Reading the servers and sending the serial, everything works fine. But when implementing the code, I start having reset in a certain function, with the following result:

```
Guru Meditation Error: Core  0 panic'ed (Unhandled debug exception). 
Debug exception reason: Stack canary watchpoint triggered (Modbus01TCP) 
Core  0 register dump:
PC      : 0x4013b3ae  PS      : 0x00060436  A0      : 0x8013c2dd  A1      : 0x3ffcffe0
A2      : 0x50c47b71  A3      : 0x3ffd0524  A4      : 0x4bfd0494  A5      : 0x3ffd00a0  
A6      : 0x00000010  A7      : 0x00007b00  A8      : 0x00000094  A9      : 0x3ffd0300
A10     : 0x00000000  A11     : 0x00000000  A12     : 0x80000020  A13     : 0x00000007  
A14     : 0x00000005  A15     : 0x00000001  SAR     : 0x00000012  EXCCAUSE: 0x00000001
EXCVADDR: 0x00000000  LBEG    : 0x4013b370  LEND    : 0x4013b3b4  LCOUNT  : 0x0000000f  


Backtrace:0x4013b3ab:0x3ffcffe00x4013c2da:0x3ffd0340 0x4013c34f:0x3ffd0360 0x4013c389:0x3ffd0380 0x40138478:0x3ffd03a0 0x401329e2:0x3ffd04a0 0x40132ad6:0x3ffd0510 0x40132b22:0x3ffd05c0 0x40132ca1:0x3ffd0640 0x4012e9d5:0x3ffd0660 0x4012eab2:0x3ffd0810 0x4012eb91:0x3ffd0830 0x4012ec61:0x3ffd08a0 0x4012c362:0x3ffd08c0 0x4012d10f:0x3ffd08f0 0x401315b5:0x3ffd0910 0x4012f1aa:0x3ffd0930 0x4012f3c0:0x3ffd0960 0x40129ed8:0x3ffd0980 0x40127951:0x3ffd0a20 0x4012798b:0x3ffd0a40 0x400e8305:0x3ffd0a60 0x400e74d0:0x3ffd0d00 0x400e764a:0x3ffd0d50 0x400e6f8a:0x3ffd0d70 0x400e7169:0x3ffd0dd0 0x400e729a:0x3ffd0e50 0x400d6e0d:0x3ffd0eb0 0x400d4cdf:0x3ffd0f00 0x401831f2:0x3ffd0f90 0x400e5066:0x3ffd0fb0 0x400e5a7d:0x3ffd0fe0 




ELF file SHA256:0000000000000000
```
   This function is called inside **void handleError(Error error, uint32_t token)**, to send errors to a Telegram bot.

**Comment by Miq1** [comment, marked as answer]
Without knowing your code I would assume the stack of the Modbus task is busted by the things you are doing in the ``onError`` handler. Mind that your ``handleError()`` code is run within that task and not in that of your main task. I see three alternatives to solve it: either just set a flag in ``handleError()`` and let the main task react on it, or do something less stack-consuming, or modify the ``ModbusClientRTU.cpp`` code where the background task is created to have a larger stack space. I would advise to do the first two, also because the more you will do in the callbacks, the more likely you will run into other such issues, like the WDT triggering etc.

**Reply by diegostf** [reply]
Using the sending of notifications in main (flag in handleError()) , the resets stopped. However, there are connection drops with the servers (timeout).
```
[WiFiClient.cpp:268] connect(): socket error on fd 49, errno: 104, "Connection reset by peer"
[ MODBUS ] Dev:2, Erro:0xEA, IP connection failed, Token: 56
```
Then I removed the code and lib from the Telegram bot. Some server disconnections still occurred.
I also removed communications with Blynk. OK, everything working.

It seems that ModbusClientTCP has conflicts with other libraries. Any way to make them work together (with Blynk, at least)?

**Reply by Miq1** [reply]
Nope, no problems with other IP communications I would be aware of.

The problem may be just the time consumed by your various communications. A Modbus TCP server has a certain timeout after which the communication is closed. It looks like the time is passing between you adding the request and its processing. The client worker task has a priority set - if other tasks have higher priority, they will prevent thr client being able to work on the request. Do you know what tasks are running on your device and with what priorities?
Mind that you have an ESP32 with two cores only and limited resources...

**Reply by diegostf** [reply]
I separated the communication routines. I await some time between them, to ensure that they were completed. They are running without fail. More tests will be necessary, but now it is smooth.
The ModbusClientTCP library runs on which core (1 or 2)? Is it possible to select one to run the task? Or this is managed by OS?

**Reply by Miq1** [reply]
Sounds like an improvement! Yes, the core for the client can be selected - please see the documentation or the ``ModbusClientTCP`` header file: ``begin()`` has an optional parameter ``coreID`` to set it.

**Comment by GeorgeIoak** [comment]
I've run across this same thing although I'm using `MODBUS01RTU` and not MODBUS01TCP`. my `handleError() is:

```
void handleError(Error error, uint32_t token)
{
    // ModbusError wraps the error code and provides a readable error message for it
    ModbusError me(error);
    LOG_WARNING("Error response: %02X - %s\n", (int)me, (const char *)me);
}
```

Currently I'm getting a lot of Timeout errors because we're working on hardware and a board could be removed for a day or two. I'm not sure but I would think once the errors are minimized this won't happen but I'm curious if there's any other advice about this so we can completely avoid getting any panics.

**Reply by Miq1** [reply]
On RTU, if you keep requesting a server that is not available, you of course will get TIMEOUT errors. There is not much to be done about it except disabling the requests in your code or selectively ignoring the error in the error handler.

**Reply by GeorgeIoak** [reply]
Sorry, yes, I understand about getting the timeout. The problem is that we're getting the panic error like the OP described. That's what I'm trying to eliminate.

**Comment by Miq1** [comment]
Aw, sorry! I did not get that from your post.

Can you provide me with the log and the output of the exception decoder from it?

**Comment by GeorgeIoak** [comment]
```[LOGFS][1680611865][WARNING]::[src/modbus.cpp:38]:Error response: E0 - Timeout

FileLogger: opening log fileFileLogger: finding next file index[53731980][E][vfs_api.cpp:105] open(): /littlefs/logs/357.log does not exist, no permits for creationFileLogger: next file index is 357
[53732000][E][vfs_api.cpp:105] open(): /littlefs/logs/357.log does not exist, no permits for creation

Guru Meditation Error: Core  0 panic'ed (Unhandled debug exception). 
Debug exception reason: Stack canary watchpoint triggered (Modbus01RTU) 
Core  0 register dump:
PC      : 0x400931fb  PS      : 0x00060036  A0      : 0x8009183c  A1      : 0x3ffda1c0  
A2      : 0x3ffbf9b8  A3      : 0xb33fffff  A4      : 0x0000abab  A5      : 0x00060023  
A6      : 0x00060023  A7      : 0x0000cdcd  A8      : 0xb33fffff  A9      : 0xffffffff  
A10     : 0x00000001  A11     : 0x00000001  A12     : 0x80085fea  A13     : 0x3ffbf0dc  
A14     : 0x007bf9b8  A15     : 0x003fffff  SAR     : 0x00000000  EXCCAUSE: 0x00000001  
EXCVADDR: 0x00000000  LBEG    : 0x4008b0d8  LEND    : 0x4008b0ee  LCOUNT  : 0xffffffff  Backtrace: 0x400931f8:0x3ffda1c0 0x40091839:0x3ffda200 0x4008fcf8:0x3ffda230 0x4008fca8:0x00060520 |<-CORRUPTEDELF file SHA256: f540945d77014bd3Rebooting...ets Jul 29 2019 12:21:46rst:0xc (SW_CPU_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)configsip: 0, SPIWP:0xeeclk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00mode:DIO, clock div:2load:0x3fff0030,len:1184load:0x40078000,len:13104load:0x40080400,len:3036entry 0x400805e4```

**Reply by Miq1** [reply]
Do you have the ESP32 Exception Decoder installed? Would be helpful to see its output as well.

The exception just tells us that the stack busted, but not the function it was in when it happened. The chances are good that it is in ``handleData()``, but we do not know for sure.

**Reply by GeorgeIoak** [reply]
No, we don't and I haven't used it before. Do you have a link on what to do or I can do some research. 

I think I've seen this 2-3 times recently but I now have 5 boards running so I can see often this happens. The MODBUS reads are on a 30 second timer right now. The 2-3 panics I think were over a 24-48 hour period.

**Reply by Miq1** [reply]
Here is the [link to the exception decoder](https://github.com/me-no-dev/EspExceptionDecoder) - installation instructions are on that page as well.

I would bet the crashes are happening in some exceptional data situations only, but we will see! 😉

**Comment by GeorgeIoak** [comment]
I did this:

C:\AquaStat>`pio device monitor -p COM14 -b 115200`

And I see this file in the directory: `platformio-device-monitor-230405-133320.log`

My platform.ini looks like this:
```
[env:esp32dev]
platform = espressif32
monitor_filters = default, esp32_exception_decoder, log2file
board = esp32dev
```

So I think it’s working although I’m not sure what the exception decoder is doing. I’m guessing that if an exception happens then something will change in the PIO window but I’m not clear about that or even if I set it up correctly, I guess time will tell.

**Reply by Miq1** [reply]
Right. The exception decoder will take the addresses that are put out in case of a crash and trace those back in the ELF firmware file. This backtrace then will be printed out to the log and tell us the state the ESP32 was in in a more readable form.

**Reply by GeorgeIoak** [reply]
I've been waiting for something to happen and as luck would have it the board has been behaving until just now. I didn't get the expected outcome as you describe so perhaps this wasn't a panic and some other reason?

What I do see in the serial log is this:
```
LOGFS][1680989232][WARNING]::[src/modbus.cpp:38]:Error response: E0 - Timeout

FileLogger: opening log file
FileLogger: finding next file index
[97648329][E][vfs_api.cpp:105] open(): /littlefs/logs/1030.log does not exist, no permits for creation
FileLogger: next file index is 1030
[97648353][E][vfs_api.cpp:105] open(): /littlefs/logs/1030.log does not exist, no permits for creation




Backtrace: 0xfffffffe:0x80091850 |<-CORRUPTED

  #0  0xfffffffe:0x80091850 in ?? ??:0




ELF file SHA256: 27685147994f3a28

Rebooting...
ets Jul 29 2019 12:21:46

rst:0xc (SW_CPU_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:1184
load:0x40078000,len:13104
load:0x40080400,len:3036
entry 0x400805e4
FileLogger: setting up
FileLogger: opening log file
FileLogger: finding next file index
[   538][E][vfs_api.cpp:105] open(): /littlefs/logs/0.log does not exist, no permits for creation
FileLogger: next file index is 0
[   562][E][vfs_api.cpp:105] open(): /littlefs/logs/0.log does not exist, no permits for creation
FileLogger: current log file size is 0
[LOGFS][1680989315][NOTICE]::[src/main.cpp:177]:[boot]: version:0.5.7
CPU0 reset reason was Software reset CPU
CPU1 reset reason was Software reset CPU
[LOGFS][1680989315][WARNING]::[src/main.cpp:184]:CPU0 reset reason was Software reset CPU
[LOGFS][1680989315][WARNING]::[src/main.cpp:186]:CPU1 reset reason was Software reset CPU
[LOG][1680989315][INFO]::[src/RelayController.cpp:7]:Relay state is 0
[LOG][1680989315][INFO]::[src/RelayController.cpp:8]:Setting relay state to LOW
mounted file system
reading config file
opened config file
...
```
Is there something else I need to do or was this just a different reset reason?

**Reply by Miq1** [reply]
Well, bad luck, even the stack itself was corrupted this time. But even that is pointing to what I am suspecting, so let me present my theory.

I suppose you are doing quite some processing in your ``handleData()`` callback, right? This function is called by the ``worker`` task of the eModbus library and thus is running with the resources of that task, not those of the main task. The ``worker`` is given a limited stack only (4kB IIRC), so when the processing will need exceptional space on the stack, the limit will be busted and unpredictable things may happen.
So your processing stays within the limits most of the time. Only if something special is found in the responses it seems to get elaborate enough to break through.

If my theory is right, you have some options:
- brute-force-ly raise the stack size in the library. I do not recommend that, though, as it will require you to remember and do the change again every time the library is upgraded. Second, if the problem is a runaway loop in your code, this will not cure the root cause and you may experience the same as now.
- move local data to global. Such the data will not occupy precious stack space.
- check your program to find potential runaway states - loops not terminating or recursions too deep.
- set some ``needsProcessing`` flag in your ``handleData()`` function and postpone the processing to a check for that flag in ``loop()`` - that runs with all available space normally.

Sorry I have no other advice at this point. Please keep me in the loop about your progress!

**Reply by GeorgeIoak** [reply]
Sorry, somehow I didn't get a notice about your post. my `handleData()` isn't doing anything special:
```
void handleData(ModbusMessage response, uint32_t token)
{
    // First value is on pos 3, after server ID, function code and length byte
    uint16_t offs = 3;
    // The device has values all as IEEE754 float32 in two consecutive registers
    // Read the requested in a loop
    for (uint8_t i = 0; i < NUM_VALUES; ++i)
    {
        offs = response.get(offs, values[i]);
        // Serial.printf("   %04X: %u\n", i + FIRST_REGISTER, values[i]);
    }
    // Signal "data is complete"
    request_time = token;
    Registry::setIHPMode(IHPMODE(values[0]));
    data_ready = true;
}
```
But of course since I've been running this experiment I haven't caught that panic that had happened several times over a few days. I'm thinking of changing in `platformio.ini` from 

`monitor_filters = default, esp32_exception_decoder, log2file`

to just

`monitor_filters = esp32_exception_decoder`

And running the tests again. Any pros/cons on trying that?

**Reply by Miq1** [reply]
I can not imagine anything done in PlatformIO could affect the running ESP32.

What does ``setIHPMode()`` do? Do you have an ``onError`` handler as well?

**Reply by GeorgeIoak** [reply]
```
void handleError(Error error, uint32_t token)
{
    // ModbusError wraps the error code and provides a readable error message for it
    ModbusError me(error);
    LOG_WARNING("Error response: %02X - %s\n", (int)me, (const char *)me);
}
```
`setIHPMode()` is just setting a value

I wasn't thinking about the change in PIO impacting the code but how the capture was working. Just thinking about what changed from before to now.

**Reply by Miq1** [reply]
Your ``LOG_WARNING()`` is innocent I suppose? 

All else I can suggest is just leaving the exception decoder on and wait for some more events. There must be a special constellation causing this - else you would see it more frequently. The context of the log preceeding the reboot may give you hints.

**Reply by GeorgeIoak** [reply]
Yeah, nothing special with the `LOG_WARNING`. I need to power down and move some stuff around so we'll see what happens after a fresh start. In a way it's good that it hasn't happened but then these kind of things tend to pop back up at the worst time if you just ignore them so I'll keep plugging along.
